### PR TITLE
Simplify type bounds.

### PR DESCRIPTION
**Pull request by paulp**
I started out looking to limit the noise from empty type
bounds, i.e. the endless repetition of

```
  class A[T >: _root_.scala.Nothing <: _root_.scala.Any]
```

This led me to be reminded of all the unnecessary and
in fact damaging overreaches which are performed during parsing.
Why should a type parameter for which no bounds are
specified be immediately encoded with this giant tree:

```
  TypeBounds(
    Select(Select(Ident(nme.ROOTPKG), tpnme.scala_), tpnme.Nothing),
    Select(Select(Ident(nme.ROOTPKG), tpnme.scala_), tpnme.Any)
  )
```

...which must then be manually recognized as empty type bounds?
Truly, this is madness.
- It deftly eliminates the possibility of recognizing
  whether the user wrote "class A[T]" or "class A[T >: Nothing]"
  or "class A[T <: Any]" or specified both bounds. The fact
  that these work out the same internally does not imply the
  information should be exterminated even before parsing completes.
- It burdens everyone who must recognize type bounds trees,
  such as this author
- It is far less efficient than the obvious encoding
- It offers literally no advantage whatsoever

Encode empty type bounds as

```
  TypeBounds(EmptyTree, EmptyTree)
```

What could be simpler.

### CODE REVIEW

**Comment by paulp** [comment]
Review by @retronym

**Comment by retronym** [comment]
Looks like a winner to me, but I might run it past the Scala meeting on Tuesday to see there are any arguments in defence of the status quo.

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala [hunk 1]
```diff
@@ -2222,16 +2222,20 @@ self =>
...
+      val lo      = bound(SUPERTYPE)
+      val hi      = bound(SUBTYPE)
+      val end     = in.offset
+      val t       = TypeBoundsTree(lo, hi)
+      val defined = List(t.hi, t.lo) filter (_.pos.isDefined)
+
+      if (defined.nonEmpty)
+        t setPos wrappingPos(defined)
+      else
+        t setPos o2p(in.offset)
```
Comment (Blaisorblade) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:2233]: What's the difference between `start`, `end` and `in.offset`? I guess `in.offset` changes, but still why are `start` and `end` saved and not used?

File: src/compiler/scala/tools/nsc/javac/JavaParsers.scala [hunk 2]
```diff
@@ -293,15 +293,8 @@ trait JavaParsers extends ast.parser.ParsersCommon with JavaScanners {
...
-          var hi: Tree = TypeTree(AnyClass.tpe)
-          if (in.token == EXTENDS) {
-            in.nextToken()
-            hi = typ()
-          } else if (in.token == SUPER) {
-            in.nextToken()
-            lo = typ()
-          }
+          val hi = if (in.token == EXTENDS) { in.nextToken() ; typ() } else EmptyTree
+          val lo = if (in.token == SUPER)   { in.nextToken() ; typ() } else EmptyTree
```
Comment (Blaisorblade) [inline src/compiler/scala/tools/nsc/javac/JavaParsers.scala:297]: Couldn't this reuse `bound` from above?

File: src/compiler/scala/tools/nsc/javac/JavaParsers.scala [hunk 3]
```diff
@@ -408,15 +401,8 @@ trait JavaParsers extends ast.parser.ParsersCommon with JavaScanners {
...
-        val hi =
-          if (in.token == EXTENDS) {
-            in.nextToken()
-            bound()
-          } else {
-            scalaDot(tpnme.Any)
-          }
-        TypeDef(Modifiers(Flags.JAVA | Flags.DEFERRED | Flags.PARAM), name, List(),
-                TypeBoundsTree(scalaDot(tpnme.Nothing), hi))
+        val hi = if (in.token == EXTENDS) { in.nextToken ; bound() } else EmptyTree
```
Comment (Blaisorblade) [inline src/compiler/scala/tools/nsc/javac/JavaParsers.scala]: Shouldn't `in.nextToken` be `in.nextToken()`, since it's not pure?

**Comment by Blaisorblade** [comment]
Just to be sure, you still canonicalize empty bounds, but only from `Typers` on, right? In general, the phases before canonicalization need to handle more equivalent or almost equivalent trees, so there are a few more risks for bugs. OTOH, GHC is proud to bring this even further for better error reporting — it performs type inference _before_ desugaring exactly for this reason, so that it never needs to show the user code that the user didn't write.

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala (hunk 1 above)
Comment (soc) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:2233]: Just a minor nitpick: Is saving three characters worth picking `lo`/`hi` over `low`/high`?

Comment (paulp) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:2233]: Or put another way, is adding three characters worth gratuitously making up different names for constructs which already have names? "lo" and "hi" are what they are called.

Comment (paulp) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:2233]: start and end are residue from the great deal of flailing it took before I managed to get this tree past the position validator. I'll remove them.

File: src/compiler/scala/tools/nsc/javac/JavaParsers.scala (hunk 2 above)
Comment (paulp) [inline src/compiler/scala/tools/nsc/javac/JavaParsers.scala:297]: "From above" is a different file. There's a great deal of duplication between them; a lot less than there once was, though.

File: src/compiler/scala/tools/nsc/javac/JavaParsers.scala (hunk 3 above)
Comment (paulp) [inline src/compiler/scala/tools/nsc/javac/JavaParsers.scala]: Yeah. Historically these were all written without () so I'm conditioned not to provide it to be consistent. But at some point they were all updated with parens.

File: src/compiler/scala/tools/nsc/javac/JavaParsers.scala (hunk 2 above)
Comment (Blaisorblade) [inline src/compiler/scala/tools/nsc/javac/JavaParsers.scala:297]: I did see it was a different file; but I see it's not worth fixing _this_ in particular, right now.

**Comment by paulp** [comment]
@Blaisorblade This is the parser. It is not performing canonicalization; if you enter "A >: Nothing <: Any" that is what will be encoded. It is simply not inventing bounds.

**Comment by Blaisorblade** [comment]
@paulp 
I think I didn't make myself clear. Reading the code more closely, I see that indeed, with your change, the parser stopped "inventing bounds", but this is now done by the typer. Now, "inventing bounds" is the canonicalization I was talking about, and it's a canonicalization since it's a pure function which turns inputs which are different but equivalent into the same output.

And my point is that whenever you change the parser to canonicalize less equivalent constructs in the source, which is something you want to do, later phases have to be more careful, since they'll have to deal with trees which were impossible before.

Probably the extra cost is still worth the more accurate error reporting, but I guess that's not for me to say since I don't know the code well enough.

**Comment by paulp** [comment]
Oh, my comment from my first version of this PR #2405 didn't make the jump. "This will impact anyone who parses TypeBoundsTrees and never expects to see an EmptyTree in there. I don't know what can be done about that beyond alerting/preparing the usual parties." So I understand it makes trees possible which weren't possible before.

**Comment by paulp** [comment]
I'm not sure anyone can call the typer a "pure function" with too straight a face. BTW you already had to deal with a potential wide range of untyped trees, which I'm sure nobody did, because they had to look for (at minimum) all of these:

```
Select(Select(Ident(nme.ROOTPKG), tpnme.scala_), tpnme.Nothing)
Select(Ident(tpnme.scala_), tpnme.Nothing)
Ident(tpnme.Nothing)
```

And you can't even know if that Nothing is the real Nothing and not some other type called Nothing until it's typed. Not to mention type aliases, etc. etc. etc. You just don't know what the bounds are until it is typed.

**Comment by Blaisorblade** [comment]
OK, so the parser was really not canonicalizing so much in a useful way.
Well, I'll leave you to work :-)

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala (hunk 1 above)
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:2233]: After losing start and end, the ='s could be aligned further left if there were a 2-letter form for "defined".  My brother is an avid Scrabble player and keeps a list of meaningful 2-letter combinations, with surprising results.

Comment (paulp) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:2233]: I wonder if that means you don't know about this... http://www.scrabble-assoc.com/ratings/state/co.html

(You have to scroll all the way down to #4.)

Comment (som-snytt) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:2233]: Then there's no excuse.

I would expect all the ='s to line up following seven-letter identifiers.

**Comment by retronym** [comment]
LGTM, and to the Scala meeting today.

**Pull request merged**